Core/RetroAchievements.cpp

```diff
@@ -605,22 +611,72 @@ void identify_and_load_callback(int result, const char *error_message, rc_client
 	g_isIdentifying = false;
 }
 
-void SetGame(const Path &path) {
+struct FileContext {
+	BlockDevice *bd;
+	int64_t seekPos;
+};
+
+static BlockDevice *g_blockDevice;
+
+void SetGame(const Path &path, FileLoader *fileLoader) {
 	if (!g_rcClient || !IsLoggedIn()) {
 		// Nothing to do.
 		return;
 	}
 
+	// TODO: Fish the block device out of the loading process somewhere else. Though, probably easier to just do it here.
+	g_blockDevice = constructBlockDevice(fileLoader);
+	if (!g_blockDevice) {
+		ERROR_LOG(ACHIEVEMENTS, "Failed to construct block device for '%s' - can't identify", path.c_str());
+		return;
+	}
+
 	rc_hash_filereader rc_filereader;
 	rc_filereader.open = [](const char *utf8Path) {
-		Path path(utf8Path);
-		FILE *f = File::OpenCFile(path, "rb");
-		return (void *)f;
+		return (void *) new FileContext{ g_blockDevice, 0 };
+	};
+	rc_filereader.seek = [](void *file_handle, int64_t offset, int origin) {
+		FileContext *ctx = (FileContext *)file_handle;
+		switch (origin) {
+		case SEEK_SET: ctx->seekPos = offset; break;
+		case SEEK_END: ctx->seekPos = ctx->bd->GetBlockSize() * ctx->bd->GetNumBlocks() + offset; break;
+		case SEEK_CUR: ctx->seekPos += offset; break;
+		default: break;
+		}
+	};
+	rc_filereader.tell = [](void *file_handle) -> int64_t {
+		return ((FileContext *)file_handle)->seekPos;
+	};
+	rc_filereader.read = [](void *file_handle, void *buffer, size_t requested_bytes) -> size_t {
+		FileContext *ctx = (FileContext *)file_handle;
+
+		int blockSize = ctx->bd->GetBlockSize();
+
+		int64_t offset = ctx->seekPos;
+		int64_t endOffset = ctx->seekPos + requested_bytes;
+		int firstBlock = offset / blockSize;
+		int afterLastBlock = (endOffset + blockSize - 1) / blockSize;
+		int numBlocks = afterLastBlock - firstBlock;
+		// This is suboptimal, but good enough since we're not doing a lot of accesses.
```

Comment:
Without looking super hard, it seems like it'd generally read aligned sectors:
https://github.com/RetroAchievements/rcheevos/blob/aa6324e886e12da0cf24dee1cf4e873571f39191/src/rhash/cdreader.c#L827C68-L827C81
https://github.com/RetroAchievements/rcheevos/blob/aa6324e886e12da0cf24dee1cf4e873571f39191/src/rhash/cdreader.c#L62

Overall for an entire ISO, might be worth checking if `offset == firstOffset && requested_bytes == numBlocks * blockSize` and just reading directly.  But probably the IO will dominate anyway...

-[Unknown]

Reply:
It only hashes a few files anyway, so there's no heavy pressure on this.

Also, the very first read it does is actually unaligned - I first tried to assume aligned sectors but that was a no-go, so I just made a general quickfix.

Reply:
Oh, it's not reading the whole ISO?  Then yeah, probably not a worry anyway.

-[Unknown]

Reply:
Yeah, it's not ideal but it's how RetroAchievements does things, and for user experience it's a good thing, of course... It hashes the eboot and paramsfo basically. Cheating through hacking other files is thus perfectly possible, but on the other hand, you can just hack the emulator instead...